core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java

```diff
@@ -222,325 +262,407 @@ synchronized void updateGlobalCheckpointOnReplica(final long globalCheckpoint) {
     }
 
     /**
-     * Get the primary context for the shard. This includes the state of the global checkpoint tracker.
+     * Called when the recovery process for a shard is ready to open the engine on the target shard. Ensures that the right data structures
+     * have been set up locally to track local checkpoint information for the shard.
      *
-     * @return the primary context
+     * @param allocationId  the allocation ID of the shard for which recovery was initiated
      */
-    synchronized PrimaryContext primaryContext() {
-        if (sealed) {
-            throw new IllegalStateException("global checkpoint tracker is sealed");
+    public synchronized void initiateTracking(final String allocationId) {
+        assert invariant();
+        assert primaryMode;
+        LocalCheckPointState lcps = localCheckpoints.get(allocationId);
+        if (lcps == null) {
+            // can happen if replica was removed from cluster but recovery process is unaware of it yet
+            throw new IllegalStateException("no local checkpoint tracking information available");
         }
-        sealed = true;
-        final ObjectLongMap<String> inSyncLocalCheckpoints = new ObjectLongHashMap<>(this.inSyncLocalCheckpoints);
-        final ObjectLongMap<String> trackingLocalCheckpoints = new ObjectLongHashMap<>(this.trackingLocalCheckpoints);
-        return new PrimaryContext(appliedClusterStateVersion, inSyncLocalCheckpoints, trackingLocalCheckpoints);
+        assert invariant();
     }
 
     /**
-     * Releases a previously acquired primary context.
+     * Marks the shard with the provided allocation ID as in-sync with the primary shard. This method will block until the local checkpoint
+     * on the specified shard advances above the current global checkpoint.
+     *
+     * @param allocationId    the allocation ID of the shard to mark as in-sync
+     * @param localCheckpoint the current local checkpoint on the shard
      */
-    synchronized void releasePrimaryContext() {
-        assert sealed;
-        sealed = false;
+    public synchronized void markAllocationIdAsInSync(final String allocationId, final long localCheckpoint) throws InterruptedException {
+        assert invariant();
+        assert primaryMode;
+        assert handOffInProgress == false;
+        LocalCheckPointState lcps = localCheckpoints.get(allocationId);
+        if (lcps == null) {
+            // can happen if replica was removed from cluster but recovery process is unaware of it yet
+            throw new IllegalStateException("no local checkpoint tracking information available");
+        }
+        assert pendingInSync.contains(allocationId) == false;
+        updateLocalCheckpoint(allocationId, lcps, localCheckpoint);
+        // if it was already in-sync (because of a previously failed recovery attempt), global checkpoint must have been
+        // stuck from advancing
+        assert !lcps.inSync || (lcps.localCheckPoint >= globalCheckpoint);
```

Comment:
can we add a message with the local checkpoint, the aid and the global checkpoint?